dkg/sync/client.go

```diff
@@ -32,168 +32,178 @@ import (
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
-	return resp, nil
+	return c.connected
 }
 
-// NewClient starts a goroutine that establishes a long lived connection to a p2p server and returns a new Client instance.
-// TODO(dhruv): call onFailure on permanent failure.
-func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hashSig []byte, onFailure func()) *Client {
-	s, err := tcpNode.NewStream(ctx, server.ID, syncProtoID)
-	if err != nil {
-		log.Error(ctx, "Open new stream with server", err)
-		ch := make(chan result, 1)
-		ch <- result{error: err}
-		close(ch)
-
-		return &Client{
-			ctx:       ctx,
-			onFailure: onFailure,
-			tcpNode:   tcpNode,
-			server:    server,
-			results:   ch,
+func (c *Client) sendMsgs(ctx context.Context, stream network.Stream) (bool, error) {
+	timer := time.NewTicker(time.Second)
+	defer timer.Stop()
+
+	first := make(chan struct{}, 1)
+	first <- struct{}{}
+
+	var shutdown bool
+
+	for {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case <-c.shutdown:
+			shutdown = true
+		case <-first:
+		case <-timer.C:
 		}
+
+		resp, err := c.sendMsg(stream, shutdown)
+		if isRelayError(err) {
+			return true, nil // Reconnect on relay errors
+		} else if err != nil {
+			return false, err
+		} else if shutdown {
+			return false, nil
+		}
+
+		rtt := time.Since(resp.SyncTimestamp.AsTime())
+		c.tcpNode.Peerstore().RecordLatency(c.peer, rtt)
 	}
+}
 
-	ctx, cancel := context.WithCancel(ctx)
-	out := make(chan result)
+func (c *Client) sendMsg(stream network.Stream, shutdown bool) (*pb.MsgSyncResponse, error) {
+	msg := &pb.MsgSync{
+		Timestamp:     timestamppb.Now(),
+		HashSignature: c.hashSig,
+		Shutdown:      shutdown,
+	}
 
-	client := &Client{
-		ctx:       ctx,
-		onFailure: onFailure,
-		tcpNode:   tcpNode,
-		server:    server,
-		results:   out,
-		stream:    s,
+	if err := writeSizedProto(stream, msg); err != nil {
+		return nil, err
 	}
 
-	go func() {
-		defer close(out)
-		defer cancel()
+	resp := new(pb.MsgSyncResponse)
+	if err := readSizedProto(stream, resp); err != nil {
+		return nil, err
+	}
 
-		for ctx.Err() == nil {
-			res := client.sendHashSignature(hashSig)
+	return resp, nil
+}
 
-			if ctx.Err() != nil {
-				return
+func (c *Client) connect(ctx context.Context, retry bool) (network.Stream, error) {
+	for {
+		s, err := c.tcpNode.NewStream(network.WithUseTransient(ctx, "sync"), c.peer, protocolID)
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		} else if err != nil {
+			if retry {
+				continue
 			}
 
-			if res.error == nil {
-				tcpNode.Peerstore().RecordLatency(server.ID, res.rtt)
-			}
+			return nil, errors.Wrap(err, "open connection")
+		}
 
-			log.Debug(ctx, "Server response", z.Any("response", res.timestamp))
+		return s, nil
+	}
+}
 
-			select {
-			case out <- res:
-			case <-ctx.Done():
-				return
-			}
-		}
-	}()
-	go func() {
-		<-ctx.Done()
-		//nolint:errcheck
-		s.Reset()
-	}()
-
-	return client
+func isRelayError(_ error) bool {
```

Comment:
nit: is this function yet to be implemented?

Reply:
yeah, should add a TODO